public/app/features/explore/state/actions.ts

```diff
@@ -466,171 +452,54 @@ export function runQueries(exploreId: ExploreId): ThunkResult<void> {
     queryState.onStreamingDataUpdated = () => {
-      const data = queryState.validateStreamsAndGetPanelData();
-      const { state, error, legacy, series } = data;
-      if (!data && !error && !legacy && !series) {
-        return;
-      }
-
-      if (state === LoadingState.Error) {
-        dispatch(processErrorResults({ exploreId, response: error, datasourceId }));
-        return;
-      }
-
-      if (state === LoadingState.Streaming) {
-        dispatch(limitMessageRate(exploreId, isLokiDataSource ? series : legacy, datasourceId));
-        return;
-      }
-
-      if (state === LoadingState.Done) {
-        dispatch(changeLoadingStateAction({ exploreId, loadingState: state }));
-      }
+      const response = queryState.validateStreamsAndGetPanelData();
+      dispatch(queryStreamUpdatedAction({ exploreId, response }));
     };
 
     dispatch(queryStartAction({ exploreId }));
 
     queryState
       .execute(datasourceInstance, transaction.options)
       .then((response: PanelData) => {
-        const { legacy, error, series } = response;
-        if (error) {
-          dispatch(processErrorResults({ exploreId, response: error, datasourceId }));
-          return;
+        if (!response.error) {
+          // Side-effect: Saving history in localstorage
+          const nextHistory = updateHistory(history, datasourceId, queries);
+          dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
         }
 
-        const latency = Date.now() - now;
-        // Side-effect: Saving history in localstorage
-        const nextHistory = updateHistory(history, datasourceId, queries);
-        dispatch(historyUpdatedAction({ exploreId, history: nextHistory }));
+        dispatch(queryEndedAction({ exploreId, response }));
+        dispatch(stateSave());
+
+        // Keep scanning for results if this was the last scanning transaction
+        if (getState().explore[exploreId].scanning) {
+          if (_.size(response.series) === 0) {
+            const range = getShiftedTimeRange(-1, getState().explore[exploreId].range);
+            dispatch(updateTime({ exploreId, absoluteRange: range }));
+            dispatch(runQueries(exploreId));
+          } else {
+            // We can stop scanning if we have a result
+            dispatch(scanStopAction({ exploreId }));
+          }
+        }
+      })
+      .catch(error => {
         dispatch(
-          processQueryResults({
+          queryEndedAction({
             exploreId,
-            latency,
-            datasourceId,
-            loadingState: LoadingState.Done,
-            series: isLokiDataSource ? series : legacy,
+            response: { error, legacy: [], series: [], request: transaction.options, state: LoadingState.Error },
           })
         );
-        dispatch(stateSave());
-      })
-      .catch(error => {
-        dispatch(processErrorResults({ exploreId, response: error, datasourceId }));
       });
   };
 }
 
-export const limitMessageRate = (
```

Comment:
Removed limitMessagRate because I couldn't get the Browser to freeze anymore for very fast throughputs during live streaming and because our current strategy is to buffer this in the data source anyway as I understand.

Reply:
Well @ryantxu in his PR so far makes sure the DS does not send more than some limit number of rows. In the panelQueryState or panelQueryRunner there is throttling on the observer function for 50ms (btw @ryantxu this only works if we send full frames right, if DS sends delta frames we would loose data?). So this may work OK right now. Kinda feel like if I wait long enough I will not have to write any code myself :D.

Reply:
[panelQueryRunner](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryRunner.ts#L224) has a 50ms throttle, panelQueryState does not.

We are punting on delta frames for now -- but yes if we just send delta without changing anything else they would drop frames.  I expect a request for 'delta' directly from a panel should maybe go through a different path.